kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/balancer/LoadBalancer.java

```diff
@@ -0,0 +1,149 @@
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+package org.kie.server.client.balancer;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+
+import org.kie.remote.common.rest.KieRemoteHttpRequest;
+import org.kie.remote.common.rest.KieRemoteHttpRequestException;
+import org.kie.server.client.balancer.impl.RandomBalancerStrategy;
+import org.kie.server.client.balancer.impl.RoundRobinBalancerStrategy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoadBalancer {
+
+    private static final Logger logger = LoggerFactory.getLogger(LoadBalancer.class);
+
+    private ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+
+    private final BalancerStrategy balancerStrategy;
+    private CopyOnWriteArraySet<String> failedEndpoints = new CopyOnWriteArraySet<String>();
+
+    protected LoadBalancer(BalancerStrategy balancerStrategy) {
+        this.balancerStrategy = balancerStrategy;
+    }
+
+    public String getUrl() throws KieRemoteHttpRequestException{
+        String selectedUrl = balancerStrategy.next();
+        logger.debug("Load balancer {} selected url '{}'", balancerStrategy, selectedUrl);
+        return selectedUrl;
+    }
+
+    public void markAsFailed(String url) {
+        failedEndpoints.add(url);
+        balancerStrategy.markAsOffline(url);
+        logger.debug("Url '{}' is marked as failed and will be considered offline by {}", url, balancerStrategy);
+    }
+
+    public void activate(String url) {
+        failedEndpoints.remove(url);
+        balancerStrategy.markAsOnline(url);
+        logger.debug("Url '{}' is marked as activated and will be considered online by {}", url, balancerStrategy);
+    }
+
+    public void close() {
+        try {
+            executorService.shutdownNow();
+        } catch (Exception e) {
+            logger.debug("Error when shutting down load balancer executor service");
+        }
+    }
+
+    /*
+     *  background operations for checking failed endpoints
+     */
+    public Future<?> checkFailedEndpoints() {
+        return executorService.submit(new CheckFailedEndpoints());
+    }
+
+    /*
+     * factory methods
+     */
+
+    public static LoadBalancer getDefault(String urls) {
+        String[] endpoints = new String[0];
+        if (urls != null) {
+            endpoints = urls.split("\\|");
```

Comment:
Could we rather define the delimiter char as a constant to prevent its repetition (see `forStrategy()` method)?